### ISSUE TITLE
"Reactor System # Advanced state system capabilities" example throws pathological stack trace

### ISSUE THREAD

**Issue by brianthelion** [saltstack/salt#43407]
### Description of Issue/Question

This [example](https://docs.saltstack.com/en/latest/topics/reactor/#advanced-state-system-capabilities
) given in the "Reactor System" docs is meant to be a toy, but because it has all the right ligature it should at least attempt to run all the files provided. Unfortunately, it does not. Instead we get the following stack trace in the master logs:

```
2017-09-07 22:41:50,964 [salt.output                                                         ][TRACE   ][31529] data = Exception occurred in runner state.orchestrate: Traceback (most recent call last):
  File "/tmp/tmp.22EsxrGXNR/lib/python3.5/site-packages/salt/client/mixins.py", line 392, in _low
    data['return'] = self.functions[fun](*args, **kwargs)
TypeError: orchestrate() missing 1 required positional argument: 'mods'
```

### Setup
Run the following script:

```bash
VENV_DIR=$(mktemp -d)
CONFIG_DIR=$(mktemp -d)
FILE_ROOT=$(mktemp -d)
ROOT_DIR=$(mktemp -d)

python3.5 -m venv $VENV_DIR
$VENV_DIR/bin/pip install salt

cat <<EOF > $CONFIG_DIR/master
root_dir: $ROOT_DIR
file_roots:
  base:
    - $FILE_ROOT
reactor:
  - myco/custom/event:
    - salt://reactor/some_event.sls
EOF

mkdir $FILE_ROOT/reactor
cat <<EOF > $FILE_ROOT/reactor/some_event.sls
# /srv/reactor/some_event.sls
invoke_orchestrate_file:
  runner.state.orchestrate:
    - mods: _orch.do_complex_thing # /srv/salt/_orch/do_complex_thing.sls
    - kwarg:
        pillar:
          event_tag: {{ tag }}
          event_data: {{ data|json() }}
EOF

mkdir $FILE_ROOT/_orch
cat <<EOF > $FILE_ROOT/_orch/do_complex_thing.sls
# /srv/salt/_orch/do_complex_thing.sls
{% set tag = salt.pillar.get('event_tag') %}
{% set data = salt.pillar.get('event_data') %}

# Pass data from the event to a custom runner function.
# The function expects a 'foo' argument.
do_first_thing:
  salt.runner:
    - name: custom_runner.custom_function
    - foo: {{ data.foo }}

# Wait for the runner to finish then send an execution to minions.
# Forward some data from the event down to the minion's state run.
do_second_thing:
  salt.state:
    - tgt: {{ data.bar }}
    - sls:
      - do_thing_on_minion
    - kwarg:
        pillar:
          baz: {{ data.baz }}
    - require:
      - salt: do_first_thing
EOF


$VENV_DIR/bin/salt-master --log-file-level=all --config-dir=$CONFIG_DIR --daemon --pid-file=$CONFIG_DIR/.pid
$VENV_DIR/bin/salt-run --log-level=all --config-dir=$CONFIG_DIR event.send 'myco/custom/event' '{}'
sleep 10

pkill -F $CONFIG_DIR/.pid
echo "###############################"
echo $ROOT_DIR
cat $ROOT_DIR/var/log/salt/master
rm -rf $CONFIG_DIR $FILE_ROOT $VENV_DIR $ROOT_DIR
echo "###############################"
```

### Versions Report
Salt Version:
           Salt: 2017.7.1
 
Dependency Versions:
           cffi: Not Installed
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
          ioflo: Not Installed
         Jinja2: 2.9.6
        libgit2: Not Installed
        libnacl: Not Installed
       M2Crypto: Not Installed
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.4.8
   mysql-python: Not Installed
      pycparser: Not Installed
       pycrypto: 2.6.1
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.5.3 (default, Jan 19 2017, 14:11:04)
   python-gnupg: Not Installed
         PyYAML: 3.12
          PyZMQ: 16.0.2
           RAET: Not Installed
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.2
            ZMQ: 4.1.6
 
System Versions:
           dist: Ubuntu 17.04 zesty
         locale: UTF-8
        machine: x86_64
        release: 4.12.0-041200-generic
         system: Linux
        version: Ubuntu 17.04 zesty

**Comment by brianthelion** [comment]
Note that the parameter that the stack trace is complaining about -- `mods` -- is present in the reactor state file. The problem behavior can also be observed with other, non-toy examples.

**Comment by gtmanfred** [comment]
This is a known issue.

If you move mods inside of the `kwarg` option, it will be fixed.

We are working on fixing this all in 2017.7.2

**Comment by gtmanfred** [comment]
https://github.com/saltstack/salt/pull/43330


This is a duplicate of https://github.com/saltstack/salt/issues/43077

**Comment by MDPolaris** [comment]
@gtmanfred Is the workaround mentioned above going to break when the fix is released? I just need to plan on whether this needs to be updated next time I upgrade versions. Thanks.

**Comment by gtmanfred** [comment]
The fix is already released in 2017.7.2, and the docs have been updated to include the suggested ways to use the reactor.

But the one above is the format that works with all past and future releases. (including 2017.7.0/1)  But the old method works with all other releases, and the new method which is better imho, works with 2017.7.2 and will work in all releases going forward

**Comment by MDPolaris** [comment]
Thanks for the update. I agree the new format is much better. I have to
wait until 2017.7.3 due to another bug, but i am looking forward to moving
to the newer format. I am glad the workaround will continue to work because
i am sure i will forget to move to the new format when i patch next and its
good that this won't break. Thanks again.
